source/code/providers/Container_ContainerLog_Class_Provider.cpp

```diff
@@ -0,0 +1,288 @@
+/* @migen@ */
+#include <MI.h>
+#include "Container_ContainerLog_Class_Provider.h"
+
+#include <map>
+#include <stdio.h>
+#include <string>
+#include <string.h>
+#include <syslog.h>
+#include <time.h>
+#include <unistd.h>
+#include <uuid/uuid.h>
+#include <vector>
+
+#include "../cjson/cJSON.h"
+#include "../dockerapi/DockerRemoteApi.h"
+#include "../dockerapi/DockerRestHelper.h"
+
+#define LASTLOGQUERYTIMEFILE "/var/opt/microsoft/docker-cimprov/state/LastLogQueryTime.txt"
+
+using namespace std;
+
+class Guid
+{
+public:
+    ///
+    /// Create a guid and represent it as string
+    ///
+    static string NewToString()
+    {
+        uuid_t uuid;
+        uuid_generate_random(uuid);
+        char s[37];
+        uuid_unparse(uuid, s);
+        return string(s);
+    }
+};
+
+MI_BEGIN_NAMESPACE
+
+class ContainerLogQuery
+{
+private:
+    ///
+    /// Get the previous time if it was stored, otherwise get the current time
+    ///
+    /// \returns Time of previous query if available
+    ///
+    static int GetPreviousTime()
+    {
+        int fileTime = time(NULL);
+        int currentTime = fileTime;
+        const char* lastQueryFile = LASTLOGQUERYTIMEFILE;
+        FILE* file = fopen(lastQueryFile, "r");
+
+        if (file)
+        {
+            fscanf(file, "%d", &fileTime);
+            fclose(file);
+
+            if (fileTime > currentTime)
+            {
+                syslog(LOG_WARNING, "The time stored in %s is more recent than the current time", lastQueryFile);
+            }
+        }
+        else
+        {
+            syslog(LOG_ERR, "Attempt in GetPreviousTime to open %s for reading failed", lastQueryFile);
```

Comment:
Purpose of this file is to get the previous time when the container logs were collected. First time when the file is not present, provider will collect the latest logs and will write the current time in this file.
Next time file will be present and will have the previous time stored.

Any kind of error is not terminating here, syslog warning is sufficient. And in case of error again provider collects the latest logs and sends them to oms.